### PR TITLE
BInterval - U fix

### PR DESCRIPTION
**Pull request by patrick-kidger**
- Fixed the `U` returned from BInterval.
- Added tests for `U`.
- Added repetitions back to `test_brownian_interval.py::test_normality_conditional`.

I've ran all the tests and all the diagnostics and everything LGTM. Sorry this took so long, getting the tests working proved surprisingly finickity.

I've not added any tests for `A`, basically because I don't see any sensible options for doing so.
We can't check consistency, i.e. is
```
\int_s^t W^1_{s,u} dW^2_u = \int_s^v W^1_{s,u} dW^2_u + \int_v^t W^1_{s,v} dW^2_u + \int_v^t W^1_{v,u} dW^2_u
```
true [or rather the antisymmetric version of this] because we're not trying to satisfy this condition anyway. Davie-Foster corrections are generated on each subinterval independently. Moreover it's not clear that generating one on one subinterval, and then calculating the other, will produce the correct behaviour on the second subinterval, as these are only approximations.

We can't check for distribution, because the distribution of `A` is unknown.

We _could_ independently implement the calculation we're already doing, and see if both approximations match. Worth doing do you think?

### CODE REVIEW

File: torchsde/_brownian/brownian_interval.py [hunk 1]
```diff
@@ -477,23 +477,18 @@ def __call__(self, ta, tb=None, return_U=False, return_A=False):
...
-                if self.have_H:
-                    H = H.clone()
-                if self.have_A:
-                    A = A.clone()
-
                 for interval in intervals[1:]:
                     Wi, Hi, Ai = interval.increment_and_levy_area()
                     if self.have_H:
-                        H += Hi + (interval.end - interval.start) * W
+                        term1 = (interval.end - interval.start) * (Hi + 0.5 * W)
```
Comment (lxuechen) [inline torchsde/_brownian/brownian_interval.py:483]: I double checked the revised formula and they look correct. 

But just to record things, could we type out the formula for aggregation in comments? Something like H_{s,t} = ... H_{s,u} ... H_{u, t} ... W_{s, u} ... 

Might be worthwhile to do the same for A.

File: torchsde/_brownian/brownian_interval.py [hunk 2]
```diff
@@ -477,23 +477,18 @@ def __call__(self, ta, tb=None, return_U=False, return_A=False):
             if len(intervals) > 1:
                 # If we have multiple intervals then add up their increments and Levy areas.
 
-                # Clone to avoid modifying the W, H, A that may exist in the cache
-                W = W.clone()
-                if self.have_H:
```
Comment (lxuechen) [inline torchsde/_brownian/brownian_interval.py:482]: Excellent! I was hoping that the clones be removed at some point.

**Comment by lxuechen** [comment]
Thanks for the fix! Apart from a few minor issues, the code looks good. 

Since the Foster approximation generates normals, we should be able to do conditioning. One potentially easy way I can think of at the moment is to just do conditioning for the standard Gaussians used to generate the additional noise. Conditioning is then just the usual Brownian bridge. We could discuss this further in a later PR.

File: tests/test_brownian_interval.py [hunk 3]
```diff
@@ -35,8 +36,9 @@
 SMALL_BATCH_SIZE = 16
 LARGE_BATCH_SIZE = 131072
 REPS = 3
+MEDIUM_REPS = 25
 LARGE_REPS = 500
-ALPHA = 0.001
+ALPHA = 0.0001
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:41]: We're now making lots and lots of tests - enough that we trigger this p-value at least once.
When I've had a bug and got a true positive from the test, then I've usually noticed very small p-values of 10^-180 etc, so this didn't seem likely to give any false negatives.

File: tests/test_brownian_interval.py [hunk 4]
```diff
@@ -163,49 +165,129 @@ def test_normality_simple(device, levy_area_approximation):
     if device == gpu and not torch.cuda.is_available():
         pytest.skip(msg="CUDA not available.")
 
-    t0_, t1_ = 0.0, 1.0
-    t0, t1 = torch.tensor([t0_, t1_], device=device)
+    t0, t1 = 0.0, 1.0
     for _ in range(REPS):
-        W = torch.tensor(npr.randn(), device=device).repeat(LARGE_BATCH_SIZE)
-        bm = torchsde.BrownianInterval(t0=t0, t1=t1, W=W, levy_area_approximation=levy_area_approximation)
+        base_W = torch.tensor(npr.randn(), device=device).repeat(LARGE_BATCH_SIZE)
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:170]: I'm not sure what you mean I'm afraid?

File: tests/test_brownian_interval.py [hunk 5]
```diff
@@ -163,49 +165,129 @@ def test_normality_simple(device, levy_area_approximation):
...
     for _ in range(REPS):
-        bm = torchsde.BrownianInterval(t0=t0, t1=t1, shape=(LARGE_BATCH_SIZE,),
+        bm = torchsde.BrownianInterval(t0=t0, t1=t1, shape=(LARGE_BATCH_SIZE,), device=device,
                                        levy_area_approximation=levy_area_approximation, pool_size=POOL_SIZE)
 
-        t_ = npr.uniform(low=t0_ + eps, high=t1_ - eps)
-        ta = npr.uniform(low=t0_ + eps, high=t1_ - eps)
-        tb = npr.uniform(low=t0_ + eps, high=t1_ - eps)
-        ta, t_, tb = sorted([t_, ta, tb])
+        for _ in range(MEDIUM_REPS):
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:207]: Just slow is all. ~10 minutes per (sub-)test.

File: tests/test_brownian_interval.py [hunk 6]
```diff
@@ -163,49 +165,129 @@ def test_normality_simple(device, levy_area_approximation):
...
+            t_ = npr.uniform(low=t0, high=t1)
+            ta = npr.uniform(low=t0, high=t1)
+            tb = npr.uniform(low=t0, high=t1)
+            ta, t_, tb = sorted([t_, ta, tb])
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:211]: Done

File: tests/test_brownian_interval.py [hunk 7]
```diff
@@ -163,49 +165,129 @@ def test_normality_simple(device, levy_area_approximation):
...
+    if device == gpu and not torch.cuda.is_available():
+        pytest.skip(msg="CUDA not available.")
+
+    t0, t1 = 0.0, 1.0
+    for _ in range(REPS):
+        bm = torchsde.BrownianInterval(t0=t0, t1=t1, shape=(LARGE_BATCH_SIZE, 2),  # 2 to have nontrivial Levy area
+                                       device=device,
+                                       levy_area_approximation=levy_area_approximation, pool_size=POOL_SIZE)
+
+        for _ in range(MEDIUM_REPS):
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:274]: Just slow.

File: tests/test_brownian_interval.py [hunk 8]
```diff
@@ -163,49 +165,129 @@ def test_normality_simple(device, levy_area_approximation):
...
+            t_ = npr.uniform(low=t0, high=t1)
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:275]: Done.

File: tests/test_brownian_interval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lxuechen) [reply tests/test_brownian_interval.py:170]: So it could be better to generate `npr.randn() * math.sqrt(t1 - t0)` just in case we change t0 or t1 in the future.

File: tests/test_brownian_interval.py [hunk 9]
```diff
@@ -163,49 +165,129 @@ def test_normality_simple(device, levy_area_approximation):
...
+
+@pytest.mark.parametrize("device", devices)
+@pytest.mark.parametrize("levy_area_approximation", ['none', 'space-time', 'davie', 'foster'])
+def test_consistency(device, levy_area_approximation):
+    if device == gpu and not torch.cuda.is_available():
+        pytest.skip(msg="CUDA not available.")
+
+    t0, t1 = 0.0, 1.0
+    for _ in range(REPS):
+        bm = torchsde.BrownianInterval(t0=t0, t1=t1, shape=(LARGE_BATCH_SIZE, 2),  # 2 to have nontrivial Levy area
```
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:270]: Ah yeah, I added that in when I was trying to test Levy area, before realising that was nontrivial. I'll take it out again.

File: torchsde/_brownian/brownian_interval.py (hunk 1 above)
Comment (patrick-kidger) [reply torchsde/_brownian/brownian_interval.py:483]: Done.

File: torchsde/_brownian/brownian_interval.py (hunk 2 above)
Comment (patrick-kidger) [reply torchsde/_brownian/brownian_interval.py:482]: What's there isn't any more efficient as the operations are now out of place instead of in-place. Just because it's marginally neater to look at, really!

File: tests/test_brownian_interval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:170]: That wouldn't change anything though - everything is conditional on this value anyway, i.e. we could hardcode it to some arbitrary number and the tests would pass/fail as before.

**Comment by patrick-kidger** [comment]
Hmm, perhaps I need to think more about Levy area. Let's defer that to a later PR though, agreed.

File: tests/test_brownian_interval.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lxuechen) [reply tests/test_brownian_interval.py:41]: Makes sense. Do you know how long it takes to run the new tests with `MEDIUM_REPS` on a typical machine disabling CUDA?  

I'm asking since long tests could slow down future dev cycles.

File: tests/test_brownian_interval.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lxuechen) [reply tests/test_brownian_interval.py:170]: Agree.

File: tests/test_brownian_interval.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:41]: I just timed it. 3 minutes in total. Not quite the zippy few seconds it was before, but not prohibitive IMO.

[the remark this answers is not in the document]
Comment (lxuechen) [reply tests/test_brownian_interval.py:41]: I see. Is there a particular advantage in doing 25 times the inner loop? Would there be a problem if just did `REPS=3` for it?

[the remark this answers is not in the document]
Comment (patrick-kidger) [reply tests/test_brownian_interval.py:41]: It's to be more sure that the complicated dependency structure of BInterval doesn't break anything. So I'd prefer to leave it as-is, but that's not a super strong preference if you really want to remove it.

[the remark this answers is not in the document]
Comment (lxuechen) [reply tests/test_brownian_interval.py:41]: I see. That makes sense, since the query times and seeds are stored. Let's keep it like that then.
